File: src/js/bsk_stack.js
class BSKStack {
  constructor(whiteRatio=0.3, blueRatio=0.4){
    this.data = []; // [{color:'W', value:...}, ...]
    this.whiteRatio = whiteRatio;
    this.blueRatio = blueRatio;
  }
  zones() {
    const n = this.data.length;
    const w = Math.floor(n * this.whiteRatio);
    const b = Math.floor(n * (this.whiteRatio + this.blueRatio));
    return [w,b];
  }
  push(v) {
    this.data.unshift({color:'W', value:v});
  }
  pop() {
    if(this.data.length === 0) return null;
    const [w,b] = this.zones();
    for(let i=0;i<Math.min(w,this.data.length);i++){
      if(this.data[i].color === 'W' || this.data[i].color === 'B'){
        return this.data.splice(i,1)[0].value;
      }
    }
    for(let i=w;i<Math.min(b,this.data.length);i++){
      if(this.data[i].color === 'B'){
        return this.data.splice(i,1)[0].value;
      }
    }
    throw new Error("Red zone locked");
  }
  updateColors(){
    const [w,b] = this.zones();
    for(let i=0;i<this.data.length;i++){
      if(i < w) this.data[i].color='W';
      else if(i < b) this.data[i].color='B';
      else this.data[i].color='R';
    }
  }
  size(){ return this.data.length; }
}
module.exports = BSKStack;

File: src/r/bsk_stack.r
BSKStack <- function(whiteRatio = 0.3, blueRatio = 0.4) {
  data <- list()
  list(
    push = function(x) {
      data <<- c(list(list(color='W', value=x)), data)
    },
    zones = function() {
      n <- length(data)
      w <- as.integer(n * whiteRatio)
      b <- as.integer(n * (whiteRatio + blueRatio))
      c(w,b)
    },
    pop = function() {
      if(length(data)==0) stop("empty")
      z <- zones()
      w <- z[1]; b <- z[2]
      for(i in seq_len(min(w,length(data)))) {
        if(data[[i]]$color %in% c('W','B')) {
          val <- data[[i]]$value
          data <<- data[-i]
          return(val)
        }
      }
      for(i in (w+1):min(b,length(data))) {
        if(i>0 && data[[i]]$color=='B') {
          val <- data[[i]]$value
          data <<- data[-i]
          return(val)
        }
      }
      stop("Red zone is locked")
    },
    update_colors = function() {
      z <- zones(); w <- z[1]; b <- z[2];
      for(i in seq_along(data)) {
        if(i <= w) data[[i]]$color <<- 'W'
        else if(i <= b) data[[i]]$color <<- 'B'
        else data[[i]]$color <<- 'R'
      }
    },
    len = function() length(data)
  )
}